blockchain/chainio.go

```diff
@@ -678,125 +619,92 @@ func serializeUtxoEntry(entry *UtxoEntry) ([]byte, error) {
+	// Bits 1-x encode height of containing transaction.
 	isCoinBase := code&0x01 != 0
-	output0Unspent := code&0x02 != 0
-	output1Unspent := code&0x04 != 0
-	numBitmapBytes := code >> 3
-	if !output0Unspent && !output1Unspent {
-		numBitmapBytes++
-	}
+	blockHeight := int32(code >> 1)
 
-	// Ensure there are enough bytes left to deserialize the unspentness
-	// bitmap.
-	if uint64(len(serialized[offset:])) < numBitmapBytes {
-		return nil, errDeserialize("unexpected end of data for " +
-			"unspentness bitmap")
+	// Decode the compressed unspent transaction output.
+	amount, pkScript, _, err := decodeCompressedTxOut(serialized[offset:])
+	if err != nil {
+		return nil, errDeserialize(fmt.Sprintf("unable to decode "+
+			"utxo: %v", err))
 	}
 
-	// Create a new utxo entry with the details deserialized above to house
-	// all of the utxos.
-	entry := newUtxoEntry(int32(version), isCoinBase, int32(blockHeight))
-
-	// Add sparse output for unspent outputs 0 and 1 as needed based on the
-	// details provided by the header code.
-	var outputIndexes []uint32
-	if output0Unspent {
-		outputIndexes = append(outputIndexes, 0)
+	entry := &UtxoEntry{
+		amount:      int64(amount),
+		pkScript:    pkScript,
+		blockHeight: blockHeight,
+		packedFlags: 0,
 	}
-	if output1Unspent {
-		outputIndexes = append(outputIndexes, 1)
+	if isCoinBase {
+		entry.packedFlags |= tfCoinBase
 	}
 
-	// Decode the unspentness bitmap adding a sparse output for each unspent
-	// output.
-	for i := uint32(0); i < uint32(numBitmapBytes); i++ {
-		unspentBits := serialized[offset]
-		for j := uint32(0); j < 8; j++ {
-			if unspentBits&0x01 != 0 {
-				// The first 2 outputs are encoded via the
-				// header code, so adjust the output number
-				// accordingly.
-				outputNum := 2 + i*8 + j
-				outputIndexes = append(outputIndexes, outputNum)
-			}
-			unspentBits >>= 1
-		}
-		offset++
-	}
+	return entry, nil
+}
 
-	// Decode and add all of the utxos.
-	for i, outputIndex := range outputIndexes {
-		// Decode the next utxo.  The script and amount fields of the
-		// utxo output are left compressed so decompression can be
-		// avoided on those that are not accessed.  This is done since
-		// it is quite common for a redeeming transaction to only
-		// reference a single utxo from a referenced transaction.
-		compAmount, compScript, bytesRead, err := decodeCompressedTxOut(
-			serialized[offset:], int32(version))
-		if err != nil {
-			return nil, errDeserialize(fmt.Sprintf("unable to "+
-				"decode utxo at index %d: %v", i, err))
-		}
-		offset += bytesRead
+// dbFetchUtxoEntryByHash attempts to find and fetch a utxo for the given hash.
+// It uses a cursor and seek to try and do this as efficiently as possible.
+//
+// When there are no entries for the provided hash, nil will be returned for the
+// both the entry and the error.
+func dbFetchUtxoEntryByHash(dbTx database.Tx, hash *chainhash.Hash) (*UtxoEntry, error) {
```

Comment:
This is really sad. Having to pass the db into `view.disconnectTransactions` is really painful. It kinda makes unifying UtxoViewpoint and a potential utxo cache a lot less seamless.